app/controllers/halls_controller.rb

```diff
@@ -1,2 +1,85 @@
 class HallsController < ApplicationController
+  # Gett all halls
+  def halls_list
+    if admin?
+      render json: Hall.all, each_serializer: HallSerializer
+    else
+      render json: { error: 'Not Allowed' }, status: :unauthorized
+    end
+  end
+
+  # Delete Hall by id
+  def delete_hall
+    if admin?
+      hall = Hall.find(delete_params[:hall_id])
+      raise ActiveRecord::RecordNotFound unless hall
+
+      check_reservations(hall)
+    else
+      render json: { error: 'Not Allowed' }, status: :unauthorized
+    end
+  end
+
+  # Create New Hall
+  def add_hall
+    if admin?
+      hall = current_user.hall.new(create_params)
+      if hall.save
+        render json: { created: hall }
+      else
+        render json: { error: 'Bad Request' }, status: :not_acceptable
+      end
+    else
+      render json: { error: 'Not Allowed' }, status: :unauthorized
+    end
+  end
+
+  # Update existant Hall by id
+  def update_hall
+    if admin?
+      hall = Hall.find(update_params[:hall_id])
+      raise ActiveRecord::RecordNotFound unless hall
+
+      hall.name = update_params[:name]
+      hall.description = update_params[:description]
+      hall.cost = update_params[:cost]
+      hall.capacity = update_params[:capacity]
+      hall.image = update_params[:image]
```

Comment:
In your `patch` request we need to send all fields to update the hall object, you can use the code below to check if the user wants to update a specific column:

```
      hall.name = update_params[:name] if update_params[:name].present?
      hall.description = update_params[:description] if update_params[:description].present?
      hall.cost = update_params[:cost] if update_params[:cost].present?
      hall.capacity = update_params[:capacity] if update_params[:capacity].present?
      hall.image = update_params[:image] if update_params[:image].present?
```
then all non-present params will keep their previous values.

Reply:
@EidHachem the params have required anyway in the permit function so in the front-end if the admin doesn't want to change a field he needs to send it as it is